Ardupilog.m

```diff
@@ -308,6 +314,53 @@ function delete(obj)
             return;
         end
 
+        function slice = getSlice(obj, slice_values, slice_type)
+        % This returns an indexed portion (a "slice") of an Ardupilog
+        % Example:
+        %    log_during_cruise = Log.getSlice([t_begin_cruise, t_end_cruise], 'TimeUS')
+        %  will return a smaller Ardupilog, only containing log data between
+        %  TimeUS values greater than t_begin_cruise and less than t_end_cruise.
+
+            % Copy all the properties, zero the number of messages
+            slice = copy(obj);
+            slice.numMsgs = 0;
+            
+            % Loop through the LogMsgGroups, slicing each one
+            for msgGroup = slice.logMsgGroups
+                % If the LogMsgGroup has been deleted, skip it
+                if ~isvalid(msgGroup)
+                    continue
+                end
+                
+                % Slice the LogMsgGroup
+                lmg_slice = slice.(msgGroup.Name).getSlice(slice_values, slice_type);
+                % If the slice is not empty, add it to the Ardupilog slice
+                if isempty(lmg_slice)
+                    delete(msgGroup)
+                else
+                    slice.(msgGroup.Name) = lmg_slice;
+                    slice.numMsgs = slice.numMsgs + size(slice.(msgGroup.Name).LineNo, 1);
+                end
+            end
+        end
+        
+    end
+    methods(Access=protected)
+        function cpObj = copyElement(obj)
```

Comment:
Matlab does, in general, a poor job of copying objects. Maybe you will like [this](https://github.com/Georacer/fault-diagnosis/blob/master/Structural%20Analysis/SubgraphGenerator/copy.m) as a better implementation of a deep copy.